modules/dnn/src/layers/cast_layer.cpp

```diff
@@ -77,26 +167,149 @@ class CastLayerImpl CV_FINAL : public CastLayer
         inputs_arr.getMatVector(inputs);
         outputs_arr.getMatVector(outputs);
 
-        CV_CheckEQ(inputs.size(), (size_t)1, "");
+        CV_Check(inputs.size(), inputs.size() == 1 || inputs.size() == 2, "Cast expects 1 (Cast) or 2 (CastLike) inputs");
         CV_CheckEQ(outputs.size(), (size_t)1, "");
 
-        if (inputs[0].depth() == outputs[0].depth())
-            inputs[0].copyTo(outputs[0]);
+        const Mat& src0 = inputs[0];
+        Mat& dst0 = outputs[0];
+
+        int runtimeTargetDepth = -1;
+        if (hasToParam)
+        {
+            runtimeTargetDepth = toCvDepth_;
+        }
         else
-            inputs[0].convertTo(outputs[0], outputs[0].depth());
+        {
+            Net::Impl* netimpl_ = getNetImpl(this);
+            if (netimpl_ && this->inputs.size() >= 2)
+            {
+                const Arg& in1_arg = this->inputs[1];
+                const ArgData& ad = netimpl_->argData(in1_arg);
+                if (ad.type >= 0)
+                    runtimeTargetDepth = CV_MAT_DEPTH(ad.type);
+            }
+            if (runtimeTargetDepth < 0 && inputs.size() >= 2 && !inputs[1].empty())
+                runtimeTargetDepth = inputs[1].depth();
+            if (runtimeTargetDepth < 0)
+                runtimeTargetDepth = src0.depth();
+        }
+        CV_CheckGE(runtimeTargetDepth, 0, "Cast: failed to resolve target data type at runtime");
+
+        int plannedDDepth = (runtimeTargetDepth == CV_16F) ? CV_32F :
+                            (runtimeTargetDepth == CV_16BF ? CV_16U : runtimeTargetDepth);
+        if (dst0.depth() != plannedDDepth)
+            dst0.create(dst0.size(), CV_MAKETYPE(plannedDDepth, src0.channels()));
+
+        Mat src = src0.isContinuous() ? src0 : src0.clone();
+        Mat dst = dst0.isContinuous() ? dst0 : dst0.clone();
+
+        const int sdepth = src.depth();
+        const int ddepth = dst.depth();
+
+        if (sdepth == runtimeTargetDepth && !(runtimeTargetDepth == CV_16F && ddepth == CV_32F))
+        {
+            src0.copyTo(dst0);
+            return;
+        }
+
+        if (runtimeTargetDepth == CV_16BF && (ddepth == CV_16BF || ddepth == CV_16U))
+        {
+            convertToBF16(src, dst);
+        }
+        else if (sdepth == CV_16BF)
+        {
+            src.convertTo(dst, ddepth);
+        }
+        else if (runtimeTargetDepth == CV_16F && ddepth == CV_32F)
+        {
+            Mat tmp16;
+            src.convertTo(tmp16, CV_16F);
+            tmp16.convertTo(dst, CV_32F);
+        }
+        else if (runtimeTargetDepth == CV_64F && ddepth != CV_64F)
+        {
+            Mat tmp64;
+            src.convertTo(tmp64, CV_64F);
+            if (ddepth == CV_32F)
+                tmp64.convertTo(dst, CV_32F);
+            else if (ddepth == CV_16U || ddepth == CV_16BF)
+            {
+                Mat tmp32;
+                tmp64.convertTo(tmp32, CV_32F);
```

Comment:
please, avoid double conversion; try to convert to the target type in one call, one pass through the tensor